cache tensor scalar_type in OperandInfo

ngimel opened the pull request:
Caches result of `scalar_type` call in TensorIterator and TensorOptions, because the call is expensive. 
This removes 120 - 150 ns of overhead (from 1.25 us to 1.12 us for out-of-place case, from 0.86 us to 0.73 us for inplace case)

Test plan: covered by existing tests
Differential Revision: D18576236

VitalyFedyunin commented on aten/src/ATen/native/TensorIterator.h:
```diff
@@ -102,11 +103,13 @@ struct CAFFE2_API OperandInfo {
-  ScalarType dtype = ScalarType::Undefined;
+  ScalarType target_dtype = ScalarType::Undefined;
+  // Caching dtype of the tensor, because scalar_type is an expensive operation
+  ScalarType tensor_dtype = ScalarType::Undefined;
```
Can you please inline comments how `target_dtype` is different from `tensor_dtype`

VitalyFedyunin reviewed (commented):
Overall looks good, but can be prone to error as `target_dtype` and `tensor_dtype` are looks very similar and can be mistyped wrong during development. I highly recommend to rename one of them.

VitalyFedyunin commented on aten/src/ATen/native/TensorIterator.h:
```diff
@@ -102,11 +103,20 @@ struct CAFFE2_API OperandInfo {
-  ScalarType dtype = ScalarType::Undefined;
-
-  bool is_type_defined() const { return dtype != ScalarType::Undefined; }
+  //target_dtype is the dtype that this tensor should have as a result of type promotion
```
Super nit: formatting mismatch other comments.

Commit:
cache tensor scalar_type in OperandInfo

Summary: Pull Request resolved: https://github.com/pytorch/pytorch/pull/30065

Differential Revision: D18576236

fbshipit-source-id: c67628d33aa6456ea0b8aeb0cf83a4fab53a9291
